### DISCUSSION TITLE
peculiar type matching error Node vs Node

### DISCUSSION THREAD

**Discussion by dave-english** [scenerystack/community#56] [Q&A] [answered]
I've adapted some JavaScript code from "States of Matter" into TypeScript, which should be returning a Node type object.
```
public static createIcon(elementType:string):Node{
  
  let icon=new phet.scenery.Node();
  switch (elementType){
      case 'NEON':
        icon = new Circle( 5, { fill: PaschenColors.NEON_COLOR } );
        break;

      case 'ARGON':
        icon = new Circle( 7, { fill: PaschenColors.ARGON_COLOR } );
        break;

      case 'HELIUM':
        icon = new Circle( 2, { fill: PaschenColors.HELIUM_COLOR } );
        break;

      case 'NITROGEN': {
        const nitrogen1 = new Circle( 5, { fill: PaschenColors.NITROGEN_COLOR } );
        const nitrogen2 = new Circle( 5, { fill: PaschenColors.NITROGEN_COLOR, left: nitrogen1.right - 4 } );
        icon = new Node( { children: [ nitrogen1, nitrogen2 ] } );
        break;
      }

      case 'HYDROGEN': {
        const hydrogen1 = new Circle( 3, { fill: PaschenColors.HYDROGEN_COLOR } );
        const hydrogen2 = new Circle( 3, { fill: PaschenColors.HYDROGEN_COLOR, left: hydrogen1.right - 4 } );
        icon = new phet.scenery.Node( { children: [ hydrogen1, hydrogen2 ] } );
        break;
      }

      default:
        assert && assert( false, 'unknown substance, unable to create icon' );
    }
    return icon;
  }
```

But when I call this function from another part of the program, to use as a parameter for another function, I'm getting an odd type mismatch error.

The specific error message is

> Argument of type 'import("/home/dougegan2/web_projects/phetsims/scenery/js/nodes/Node").default' is not assignable to parameter of type 'Node'.
>   Type 'Node' is missing the following properties from type 'Node': baseURI, childNodes, firstChild, isConnected, and 44 more.ts(2345)
> const nitrogenSelectorInfo: {
>     label: any;
>     icon: Node;
> }
> 

Without posting the entire rest of the code, these are the significant lines which trigger the mismatch error in VSCode
```
import { Node, NodeOptions, Rectangle, Text, VBox } from '../../../../scenery/js/imports.js';

    const nitrogenSelectorInfo = {  label: nitrogenText, icon:PaschenMoleculesControlPanel.createIcon( 'NITROGEN' ) };    
//some other code omitted
      {
        value: 'NITROGEN',
        createNode: () => new SubstanceSelectorNode( nitrogenSelectorInfo.label, nitrogenSelectorInfo.icon, selectorWidth ),
        tandemName: 'nitrogenRadioButton'
      }
//the VSCode throws the mismatch error for nitrogenSelectorInfo.icon in the same line as createNode.

```

Weirdly, there are no such mismatch errors in the original Javascript implementation from "States of Matter", although I've rewritten the code dramatically enough that I could have easily inserted my own errors.

I appreciate everyone on here who has been so patiently answering my questions.

**Comment by pixelzoom** [comment]
Unfortunately, scenery uses the same class names as some classes in the built-in JavaScript DOM API.  This includes [Node](https://developer.mozilla.org/en-US/docs/Web/API/Node) and [Text](https://developer.mozilla.org/en-US/docs/Web/API/Text), and there may be others that I'm forgetting.   So the source of this problem is almost certainly a missing import of scenery's Node somewhere in your code.

**Reply by dave-english** [reply]
Thanks.  Yes, I've been befuddled by these types of errors a couple of times.
Once the proper import statement is in the code, is there anything else necessary to disambiguate between the built in "Text" and the scenery "Text" (or Node and Node)
The reason I ask is that I was able to address all the type-mismatch warnings in my code, but was only able to do this by replacing 
```
    const neonText = new Text(
      neonString,
      merge( { tandem: radioButtonGroupTandem.createTandem( 'neonText' ) }, textOptions )
    );

```
which worked fine in the JavaScript file, now having to replace with 
```
     const neonText = new phet.scenery.Text(
      neonString,
      merge( { tandem: radioButtonGroupTandem.createTandem( 'neonText' ) }, textOptions )
    );

```

and I had to write similar constructs for the new Node() vs new phet.scenery.Node()

I had expected (wrongly, perhaps) that the import declaration would make this unnecessary, especially since it didn't appear to be necessary in the Javascript files from States of Matter.

**Reply by pixelzoom** [reply, marked as answer]
Don't feel bad - this Node/Text import problem still bites me at least once a week.

Your expectation is correct that the `import` should make qualifying like `phet.scenery.Text` unnecessary.  There must be something else going on here.  Do you have a more complete example, perhaps checked into GitHub?

**Reply by dave-english** [reply]
Thank you. I have just posted an update to my personal GitHub repository, and will send you an invite to look at it when I have a chance, and if you're willing.

This has been an interesting summer project for me, learning the SceneryStack library and improving my JavaScript skills.  I'm starting back to my regular job as a HS chemistry teacher next week.  I've been coding a simulation of the Paschen Experiment (a study of the minimum Voltage to produce an electric arc through a gas, which depends on factors such as the type of gas, distance between electrodes, and gas pressure). The project was inspired by a suggestion by Oak Nelson, a Plasma scientist at Columbia University, who has been giving me encouragement.

There are still a number of features I'd like to add to the Sim, but it is far enough along that I might write up what I've learned and accomplished for a post in the Show and Tell section of this forum.

**Comment by pixelzoom** [comment]
Thanks @dave-english, I received your GitHub invite.  I'm out until 9/17, so perhaps someone from the PhET team can respond and have a look.  If not, I'll have a look when I return.

**Comment by samreid** [comment]
Hello @dave-english, I'm interested to see your Show and Tell when it is ready! Also, I'm happy to take a look if there is still problematic behavior regarding the imports or anything.
